src/server/generic_family.cc

```diff
@@ -243,161 +253,189 @@ OpResult<RestoreArgs> RestoreArgs::TryFrom(const CmdArgList& args) {
-      pv_ = std::move(it->second);
-      it->second.SetExpire(has_expire);
-    }
+OpStatus Renamer::DelSrc(Transaction* t, EngineShard* shard) {
+  auto res = shard->db_slice().FindMutable(t->GetDbContext(), src_key_);
+  auto& it = res.it;
 
-    res.post_updater.Run();
-    CHECK(es->db_slice().Del(t->GetDbIndex(), it));  // delete the entry with empty value in it.
-    if (es->journal()) {
-      RecordJournal(t->GetOpArgs(es), "DEL", ArgSlice{src_res_.key}, 2);
-    }
+  CHECK(IsValid(it));
+
+  DVLOG(1) << "Rename: removing the key '" << src_key_;
+
+  res.post_updater.Run();
+  CHECK(shard->db_slice().Del(t->GetDbIndex(), it));
+  if (shard->journal()) {
+    RecordJournal(t->GetOpArgs(shard), "DEL", ArgSlice{src_key_}, 2);
   }
 
   return OpStatus::OK;
 }
 
-OpStatus Renamer::UpdateDest(Transaction* t, EngineShard* es) {
-  if (es->shard_id() != src_sid_) {
-    auto& db_slice = es->db_slice();
-    string_view dest_key = dest_res_.key;
-    auto res = db_slice.FindMutable(t->GetDbContext(), dest_key);
-    auto& dest_it = res.it;
-    bool is_prior_list = false;
-
-    if (IsValid(dest_it)) {
-      bool has_expire = dest_it->second.HasExpire();
-      is_prior_list = dest_it->second.ObjType() == OBJ_LIST;
-
-      if (src_res_.ref_val.ObjType() == OBJ_STRING) {
-        dest_it->second.SetString(str_val_);
-      } else {
-        dest_it->second = std::move(pv_);
-      }
-      dest_it->second.SetExpire(has_expire);  // preserve expire flag.
-      db_slice.UpdateExpire(t->GetDbIndex(), dest_it, src_res_.expire_ts);
-    } else {
-      if (src_res_.ref_val.ObjType() == OBJ_STRING) {
-        pv_.SetString(str_val_);
-      }
-      auto op_res =
-          db_slice.AddNew(t->GetDbContext(), dest_key, std::move(pv_), src_res_.expire_ts);
-      RETURN_ON_BAD_STATUS(op_res);
-      res = std::move(*op_res);
-    }
+OpStatus Renamer::DeserializeDest(Transaction* t, EngineShard* shard) {
+  auto& db_slice = shard->db_slice();
+  auto res = db_slice.FindMutable(t->GetDbContext(), dest_key_);
 
-    dest_it->first.SetSticky(src_res_.sticky);
+  auto& dest_it = res.it;
 
-    if (!is_prior_list && dest_it->second.ObjType() == OBJ_LIST && es->blocking_controller()) {
-      es->blocking_controller()->AwakeWatched(t->GetDbIndex(), dest_key);
-    }
-    if (es->journal()) {
-      OpArgs op_args = t->GetOpArgs(es);
-      string scratch;
-      // todo insert under multi exec
-      RecordJournal(op_args, "SET"sv, ArgSlice{dest_key, dest_it->second.GetSlice(&scratch)}, 2,
-                    true);
-      if (dest_it->first.IsSticky()) {
-        RecordJournal(op_args, "STICK"sv, ArgSlice{dest_key}, 2, true);
-      }
-      if (dest_it->second.HasExpire()) {
-        auto time = absl::StrCat(src_res_.expire_ts);
-        RecordJournal(op_args, "PEXPIREAT"sv, ArgSlice{dest_key, time}, 2, true);
-      }
-      RecordJournalFinish(op_args, 2);
+  int rdb_version = 0;
+  CHECK(VerifyFooter(serialized_value_.value, &rdb_version));
```

Comment:
Lets not check fail here, this can be a dcheck, and we can print error on opt.
Also I would move this to the SerializeSrc function as if we have error here we will print and return error to user but not delete the source value if this is in SerializeSrc

Reply:
Notice that `VerifyFooter` function initializes the `rdb_version` that is used later. So in the case of dcheck, the function will not be called and the value will not be initialized in opt. But yes, we can send a message to the user in case of an error.

Actually, it is not good that the function modifies the argument passed by the pointer, as in the C code. I'll refactore that.

Reply:
sounds good, looks like we can add rdb_version to SerializedValue struct and call it from SerializeSrc